src/unix/lwt_io.ml

```diff
@@ -1473,21 +1473,40 @@ let create_temp_dir
 (* This is likely VERY slow for directories with many files. That is probably
    best addressed by switching to blocking calls run inside a worker thread,
    i.e. with Lwt_preemptive. *)
-let rec delete_recursively directory =
-  Lwt_unix.files_of_directory directory
-  |> Lwt_stream.iter_s begin fun entry ->
-    if entry = Filename.current_dir_name ||
-       entry = Filename.parent_dir_name then
-      Lwt.return ()
+let delete_recursively directory =
+  let win32_unlink fn =
+    try Lwt_unix.unlink fn with
+    | Unix.Unix_error (Unix.EACCES, _, _) as e -> (
+      try
+        (* Try removing the read-only attribute *)
+        Lwt_unix.chmod fn 0o666 >>= fun () ->
+        Lwt_unix.unlink fn
+      with
+      | _ -> raise e)
+  in
+  let unlink =
+    if Sys.win32 then
```

Comment:
You probably want to lift up the `Sys.win32` check to be in a toplevel function, as then the inliner can optimise it as a constant.  This may not happen here as you're allocating the `win32_unlink` closure just above (but need to check the lambda). Not a big optimisation given the delete file time will dominate, but it's worth keeping OS-specific functions near the top of the program logic anyway to make it more obvious what's going on.

Reply:
Well that’s on me, I was supposing that getting the function closer to its callpoint would help inlining or branch prediction in some way.

Reply:
Applied your suggestion, thanks!